### PR TITLE
New lint `bytes_count_to_len`

### LINKED ISSUE #8083: Lint suggesting to replace `str::bytes().count()` with `str::len()`
### What it does

It checks for `str::bytes().count()` and suggests replacing it with `str::len()` that does the same but shorter and possibly more performant.

Extra: can something like this be done for https://doc.rust-lang.org/std/io/trait.Read.html#method.bytes too?

### Lint Name

bytes_count_to_len

### Category

complexity

### Advantage

- Might be more performant
- Shorter
- Possibly easier to read

### Drawbacks

_No response_

### Example

```rust
fn main() {
    "hello".bytes().count();
}
```

Could be written as:

```rust
fn main() {
    "hello".len();
}
```

### PR DESCRIPTION
**Pull request by c-rus**
Fixes #8083 

changelog:
• adds new lint ``[`bytes_count_to_len`]`` to consider replacing `.bytes().count()` with `.len()`

### CODE REVIEW

File: clippy_lints/src/bytes_count_to_len.rs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    /// "hello".bytes().count();
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// "hello".len();
+    /// ```
+    #[clippy::version = "1.60.0"]
+    pub BYTES_COUNT_TO_LEN,
+    complexity,
+    "Using bytest().count() when len() performs the same functionality"
```
Comment (Manishearth) [inline clippy_lints/src/bytes_count_to_len.rs:27]:
```suggestion
    "Using bytes().count() when len() performs the same functionality"
```

File: clippy_lints/src/bytes_count_to_len.rs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+declare_lint_pass!(BytesCountToLen => [BYTES_COUNT_TO_LEN]);
+
+impl<'tcx> LateLintPass<'tcx> for BytesCountToLen {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx hir::Expr<'_>) {
+        if_chain! {
+            //check for method call called "count"
+            if let hir::ExprKind::MethodCall(count_path, count_args, _) = &expr.kind;
+            if count_path.ident.name == rustc_span::sym::count;
```
Comment (Manishearth) [inline clippy_lints/src/bytes_count_to_len.rs:37]: We should use `match_def_path` here and below

And we should ensure the type is `str`

**Review by Manishearth — CHANGES_REQUESTED** [review]
Good start, some suggestions

**Comment by giraffate** [comment]
@c-rus ping from triage. Can you have any question to work on this?

**Comment by kyoto7250** [comment]
I'm interested in this PR.
Can I take over and create another PR?

**Comment by giraffate** [comment]
Yes, thanks!

**Comment by kyoto7250** [comment]
I opened a PR that took over.

https://github.com/rust-lang/rust-clippy/pull/8711

**Pull request closed without merging**